FIX Major bugfix in python refinement

caspervdw opened the pull request:
This removes any discrepancy between the python and numba engines:

 - the neighborhood should be calculated from rounded coordinates
 - the ndimage interpolation makes the code different from the numba engine and appears to add nothing in my tests

Fixes #376

tacaswell commented:
:+1:  makes sense to me.  @danielballan or @nkeim should merge this as I am only passingly familiar with this code.

caspervdw commented:
In view of the API change this will give, some more testing data below. Apparently, the numba code was made more accurate by #242 without us realizing it. Now the python code is also fixed.

I guess this pure python refine could use some streamlining, but that is for another time.

this PR:
![refine_trackpy_tests_pr](https://cloud.githubusercontent.com/assets/8937919/16050252/3b50fffe-325b-11e6-97f6-6bc3ef0bb654.png)

v0.3.0:
![refine_trackpy_tests_master](https://cloud.githubusercontent.com/assets/8937919/16050224/2707339c-325b-11e6-9df0-2a4ea3db06c8.png)

v0.2.4:
![refine_trackpy_tests_v0 2 4](https://cloud.githubusercontent.com/assets/8937919/16050321/842db8d4-325b-11e6-99ad-4c4ff7d81064.png)

nkeim commented:
@caspervdw This is great!! Thank you for tracking this down.

Your test results suggest that there are reliable upper bounds on positional error. Do you think this would work as an automated test? At this point I think that creating an issue would suffice.

caspervdw commented:
I was thinking the same. There is a test already, `test_ep`. I wonder why it passed. Maybe we need to sharpen the tolerances.

danielballan commented:
Probably. If you have the code that generated those plots handy, maybe we should start a stash of "human-in-the-loop" tests to be run before releases and on PRs that touch critical code.

caspervdw commented:
Shall we add this to v0.3.1?

danielballan commented:
Is the signal-to-noise ratio on the plots actually a noise-to-signal ratio? The trends are the opposite of what I would expect, unless I am confused.

My vote is that this is a bug fix, not an API change, and thus can go into v0.3.1 (not v0.4.0).

tacaswell commented:
I am also :+1: on this being a bugfix, not an API change.

On Sun, Jun 19, 2016 at 5:49 PM Dan Allan <notifications@github.com> wrote:

> Is the signal-to-noise ratio on the plots actually a noise-to-signal
> ratio? The trends are the opposite of what I would expect, unless I am
> confused.
>
> My vote is that this is a bug fix, not an API change, and thus can go into
> v0.3.1 (not v0.4.0).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/soft-matter/trackpy/pull/377#issuecomment-227022412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAMMhbtg8vxu52_vr2HBqXnIi4tsy-_tks5qNblxgaJpZM4I0slI>
> .
>

caspervdw commented:
@danielballan  Yes it should be noise-to-signal ratio... I will have a quick look at `test_ep` today so that we can merge soon.

caspervdw commented:
Done, ready to merge

nkeim commented:
Sorry for taking a closer look so late. It looks like this removes the shift code that was the principal difference between the python and numba engines, going all the way back to the beginning. I had always thought that this made the python engine slightly superior! But the data don't lie.

caspervdw commented:
I'll quickly run a test today to check if there is a difference @nkeim

caspervdw commented:
Apparently, it really doesn't make a difference whether you interpolate or not:

![refine_trackpy_tests](https://cloud.githubusercontent.com/assets/8937919/16278212/552eae24-38b6-11e6-81d3-337c47ad2694.png)

Or, the differences with numba (negative = better, positive = worse):
![refine_trackpy_iterpolation_tests](https://cloud.githubusercontent.com/assets/8937919/16278261/8b34ffbe-38b6-11e6-883f-66137da57616.png)

danielballan commented:
Interesting! I had wondered about that too. It's in the original Crocker-Grier implementation but, IIRC correctly, not in Blair's code. I wonder if Crocker tried it without interpolation first or what motivated that in the first place.

danielballan commented on trackpy/feature.py:
```diff
@@ -361,18 +362,19 @@ def _refine(raw_image, image, radius, coords, max_iterations,
+                #                              order=2, mode='constant', cval=0)
+                # new_coord = coord + off_center
+                # # Disallow any whole-pixels moves on future iterations.
+                # allow_moves = False
```
All references to `allow_moves` can be removed. It will always be `True`.

caspervdw replied:
I'll look at it and refactor. Also the `shift_thresh` and `break_thresh` will be the same thing: if the code does not shift, it will break. Calculating the CoM without shifting makes no sense because the outcome will not change.

caspervdw commented:
Done the refactor at @danielballan suggestion. Because the interpolation went out, the refinement loop could be simplified substantially: `break_thresh` and  `shift_thresh` both resulted in stopping of the refinement loop. `break_thresh` basically did not have a function anymore. I dropped it as parameter for `refine`, I have added a deprecation warning.

I ran the tests again and the performance is better than before. Maybe I solved another bug: the `off_center` might have been subtracted one time too much, previously. But I am not sure of the origin of this difference. It is for sure much better at low noise values.

![refine_trackpy_tests_refactored](https://cloud.githubusercontent.com/assets/8937919/16295573/04cabcb2-3928-11e6-8e48-58234bfcf6c6.png)

caspervdw commented:
The test should be fixed now, the resulting tracking errors are now lower, agreeing much better (within 10%) with the theoretical ones.

caspervdw commented:
Could anyone check this and merge? The amount of code that is changed is substantial, but also fairly straightforward. I think that this increase of accuracy should be released asap so that people can make use of it.

Potentially as a v0.4, not v0.3.1? What do you think? This PR will result in different refined coordinates.

nkeim commented:
I suspect that the "python" engine is used by only a small minority. There is zero downside to this PR except that it changes output. And putting out a release is a fair amount of work. So I vote to include it in 0.3.1. I would like to focus on #361 (and maybe #333) for v0.4.

nkeim commented:
I've looked at the code and am happy with it. If I had more time I would test this with my application (1D feature finding). But if there is a problem with that very rare use case, I will find it soon enough.

So I also want to merge if we are OK with this being in v0.3.1.

danielballan commented:
Yep, this is a bugfix and can go into v0.3.1. Merging!
